### ISSUE TITLE
[RFC] Routing Replay (R3) for vLLM Rollout Engine

### ISSUE THREAD

**Issue by CalvinXKY** [vllm-project/vime#32]
**Status:** In progress  
**Target model:** Qwen3-30B-A3B (48 MoE layers, `moe_router_topk=8`)  
**Rollout backend:** vLLM (via slime `vllm_rollout`)

---

## Summary

**Routing Replay (R3)** records MoE expert routing decisions during rollout and replays them during Megatron training so the training forward pass uses the same expert assignments as inference.

SGLang already supports R3 end-to-end (`--enable-return-routed-experts` + `meta_info.routed_experts`). vLLM exposes routing internally when launched with `--enable-return-routed-experts`, but the token-only `/inference/v1/generate` path used by slime’s `vllm_rollout` did not surface those fields on the HTTP response, and the client did not merge prompt vs. generation segments into the tensor shape Megatron expects.

This RFC defines a **phased plan** to bring vLLM rollout to R3 parity with SGLang, starting with the **currently deployed vLLM 0.21.0** stack, then upgrading to newer vLLM releases.

---

## Phased Roadmap

| Phase | Scope | Goal |
|-------|--------|------|
| **Phase 1 (current)** | **vLLM 0.21.0** | Patch HTTP, align Megatron tensor contract, update vime client/engine launcher, validate decoupled then colocate |
| **Phase 2 (follow-up)** | **Latest vLLM** | Rebase or drop patch, re-validate API payload & launch flags, extend CI matrix |

Phase 1 is a **version-pinned adapter**. Phase 2 is a separate upgrade track once Phase 1 is stable in training.

---

## Background

When `--use-rollout-routing-replay` is enabled:

1. The rollout engine returns per-token expert indices for every MoE layer.
2. `vllm_rollout` / `sglang_rollout` populate `sample.rollout_routed_experts`.
3. Megatron actor consumes `rollout_routed_experts` during the training step.

Megatron requires:

```text
rollout_routed_experts.shape == (len(sample.tokens) - 1, num_layers, moe_router_topk)
dtype: int32
```

For Qwen3-30B-A3B: `(seq_len - 1, 48, 8)`.

**SGLang (reference):** single base64 int32 buffer in `meta_info.routed_experts`, reshaped directly.

**vLLM 0.21.0 (engine internal):** routing split into:

| Segment | HTTP field | Role |
|---------|------------|------|
| Prompt | `prompt_routed_experts` (response top-level) | Prompt-phase routing |
| Generation | `choices[].routed_experts` | Decode-phase routing |

The slime client must **concatenate** both segments before assigning `sample.rollout_routed_experts`.

---

### Phase 1 (vLLM 0.21.0)

- [ ] Expose `prompt_routed_experts` and `choices[].routed_experts` on `/inference/v1/generate`.
- [ ] Parse and merge routing in `vllm_rollout._apply_vllm_routed_experts` to match SGLang/Megatron contract.
- [ ] Launch vLLM with R3-compatible flags from `vllm_engine.launch_server_process`.
- [ ] Validate **decoupled** train/rollout layout first, then **colocate** layout.
- [ ] Add unit tests and integration-style smoke tests aligned with existing slime CI patterns.

### Phase 2 (latest vLLM — future)

- [ ] Re-evaluate whether HTTP patch is still needed.
- [ ] Re-run cross-engine diff and both deployment-mode smokes.
- [ ] Pin supported vLLM version range in docs/CI.

## Non-Goals

- Changing Megatron routing-replay kernel logic (works with SGLang today).
- Bit-exact routing match across two independent inference runs (shape/dtype contract is the gate).
- SGLang backend changes (SGLang remains the reference implementation).

---

### Step 1 — Patch vLLM 0.21.0 HTTP layer

**Problem:** Engine computes routing, but `GenerateResponse` on `/inference/v1/generate` omitted routing fields.

**Change:** Apply `vime/docker/patch/latest/vllm.patch`:

- `GenerateResponseChoice.routed_experts` — generation segment (nested int lists).
- `GenerateResponse.prompt_routed_experts` — prompt segment.
- Wire fields in `ServingTokens` from engine outputs.

**Apply methods:**

| Method | Location |
|--------|----------|
| Runtime | `run_script/apply_vllm_generate_routed_experts_patch.sh` |
| Image build | `vime/docker/Dockerfile` |

**Required server flags (0.21.0):**

```bash
VLLM_SERVER_DEV_MODE=1 python -m vllm.entrypoints.openai.api_server \
  --enable-return-routed-experts \
  --no-async-scheduling \
  # MoE multi-GPU (opt-in via vime CLI):
  --enable-expert-parallel \
  --expert-placement-strategy linear
```

> Routed-experts replay is **incompatible with async scheduling** on vLLM 0.21.x.

**Quick HTTP gate:**

```bash
apply_vllm_generate_routed_experts_patch.sh --curl
```

Validates `/v1/completions` and `/inference/v1/generate` both expose routing fields.

---

### Step 2 — Align Megatron-facing format

Megatron actor (`slime/backends/megatron_utils/actor.py`) asserts `rollout_routed_experts` is present when `--use-rollout-routing-replay` is set.

**vLLM client logic** (`vllm_rollout._apply_vllm_routed_experts`):

1. Read `output["prompt_routed_experts"]` and `choice["routed_experts"]`.
2. Support **nested lists** (0.21.0 patched HTTP) and **base64 npy** (legacy path).
3. `np.concatenate` along axis 0.
4. Validate `(len(tokens)-1, num_layers, moe_router_topk)`; assign to `sample.rollout_routed_experts`.

No Megatron changes required once the rollout tensor contract matches SGLang.

---

### Step 3 — Cross-engine validation (SGLang vs vLLM)

Before merging client changes, use offline tooling under `verfiy_rollout/`:

| Artifact | Purpose |
|----------|---------|
| `rollout_routing_extract.py` | Mirrors slime extraction for both engines |
| `fetch_curl_responses.sh` | Same prompt token IDs → SGLang + vLLM JSON |
| `compare_curl_rollout.py` | Shape parity check |
| `run_compare_on_server.sh` | One-shot runner |

**Pass criterion:** both engines produce `(len(tokens)-1, 48, 8)`. `exact_match=False` across two runs is acceptable.

---

#### 4.1 `vime/slime/rollout/vllm_rollout.py`

- `_vllm_routed_experts_payload_to_array()` — list or base64 npy.
- `_apply_vllm_routed_experts()` — merge prompt + generation segments.
- `generate()` passes full `output` to `_apply_vllm_routed_experts`.

#### 4.2 `vime/slime/backends/vllm_utils/vllm_engine.py`

When `use_rollout_routing_replay=True`:

```python
cmd += ["--enable-return-routed-experts"]
if not user_overrode("vllm_async_scheduling"):
    cmd += ["--no-async-scheduling"]
if args.vllm_enable_expert_parallel:
    cmd += ["--enable-expert-parallel", "--expert-placement-strategy", "linear"]
```

Expert parallel is **opt-in** via `--vllm-enable-expert-parallel`.

#### 4.3 Tests

| Test | Coverage |
|------|----------|
| `tests/unit/rollout/test_vllm_rollout.py` | Merge logic, shape validation, mocked `generate()` |
| `tests/test_vllm_generate_endpoint.py` | Live vLLM server smoke |

---

## Verification Plan (ordered)

Validation proceeds in **two deployment modes**, from simpler to more complex.

### Gate A — Unit & HTTP (no full train loop)

1. Unit tests: `pytest tests/unit/rollout/test_vllm_rollout.py`
2. Patch + HTTP: `apply_vllm_generate_routed_experts_patch.sh --curl`
3. Cross-engine offline diff: `verfiy_rollout/run_compare_on_server.sh`

---

### Gate B — Decoupled mode (train / rollout GPU split) ✅ first

**Layout:** Megatron actor and rollout engine use **disjoint GPU sets** (no `--colocate`).

**Reference script:** `run_script/qwen3-30B-A3B-4t4i.sh`

Example topology (8 GPUs):

- Train: 4 GPUs (`--actor-num-gpus-per-node 4`, EP=4)
- Rollout: 4 GPUs (`--rollout-num-gpus 4`, `--rollout-num-gpus-per-engine 4`, vLLM TP=4)

Key flags:

```bash
--use-rollout-routing-replay
--vllm-enable-expert-parallel
--vllm-gpu-memory-utilization 0.70
--vllm-max-model-len 8192
# no --colocate
```

**Smoke (rollout-only):**

```bash
bash run_script/verify_routing_replay_qwen3_30b.sh
# VERIFY_MODE=rollout  → debug-rollout-only, num-rollout 1
# validates rollout_0.pt shape via verify_rollout_routed_experts_pt.py
```

**Full integration (recommended before colocate):**

- Short `num-rollout` run with `qwen3-30B-A3B-4t4i.sh` (or equivalent).
- Confirm no Megatron assertion: `rollout_routed_experts is required in rollout_data`.
- Confirm training step completes with injected routing.

**Pass criteria (decoupled):**

- [ ] `rollout_routed_experts.shape == (len(tokens)-1, 48, 8)` on rollout samples
- [ ] Weight sync + rollout + at least one train step succeed
- [ ] No OOM/regression from `--no-async-scheduling` + EP flags

---

### Gate C — Colocate mode (shared GPUs) ✅ second

**Layout:** Actor and rollout **share the same GPU pool** (`--colocate`).

**Reference test (canonical slime R3 colocate recipe):**  
`vime/tests/test_qwen3_30B_A3B_r3.py`

This test defines the **target flag bundle** for colocate R3 on Qwen3-30B-A3B:

| Area | Key settings from `test_qwen3_30B_A3B_r3.py` |
|------|-----------------------------------------------|
| R3 | `--use-rollout-routing-replay` |
| Layout | `--colocate`, `--actor-num-gpus-per-node 8`, 8 GPUs total |
| Train parallel | TP=4, SP, CP=2, EP=8 |
| Rollout | `--rollout-num-gpus-per-engine 8` (SGLang in reference test) |
| Algo | GSPO + TIS, `num-rollout 3`, `--ci-test` |
| MoE | DeepEP optional (`SLIME_TEST_USE_DEEPEP`) |

**Phase 1 deliverable:** a **vLLM-backed variant** of this test (same topology and hyperparameters, swap SGLang rollout args for vLLM):

```python
# Planned: tests/test_qwen3_30B_A3B_r3_vllm.py
# Structure mirrors test_qwen3_30B_A3B_r3.py:
#   - same prepare() / perf_args / grpo_args / misc_args (--colocate)
#   - replace sglang_args with vllm_args:
#       --rollout-num-gpus-per-engine 8   # or TP=4 × DP=2 per cluster layout
#       --vllm-enable-expert-parallel
#       --vllm-gpu-memory-utilization ...
#       --vllm-max-model-len ...
#   - keep --use-rollout-routing-replay and --ci-test
```

Until that test lands, manual colocate smoke should **copy argument groups verbatim** from `test_qwen3_30B_A3B_r3.py` and only replace the rollout-backend section with vLLM equivalents.

**Pass criteria (colocate):**

- [ ] Same R3 tensor contract under `--colocate`
- [ ] `offload_train` / `offload_rollout` sleep-wake cycle does not drop routing fields
- [ ] `num-rollout ≥ 3` CI-style run passes (matching reference test depth)
- [ ] No regression vs decoupled baseline (routing shape, train step, memory)

---

## Execution Script Roadmap

| Stage | Script / test | Mode |
|-------|---------------|------|
| Patch + HTTP | `run_script/apply_vllm_generate_routed_experts_patch.sh` | Either |
| Offline diff | `verfiy_rollout/run_compare_on_server.sh` | Either |
| Rollout smoke | `run_script/verify_routing_replay_qwen3_30b.sh` | Decoupled |
| Production decoupled | `run_script/qwen3-30B-A3B-4t4i.sh` | Decoupled |
| **CI colocate (target)** | **`tests/test_qwen3_30B_A3B_r3.py` → vLLM port** | Colocate |

Future shell wrappers should **not invent new flag sets**; they should derive from `test_qwen3_30B_A3B_r3.py` (colocate) or `qwen3-30B-A3B-4t4i.sh` (decoupled).

---

## Known Limitations (Phase 1 / vLLM 0.21.0)

| Item | Detail |
|------|--------|
| Version pin | Adapter targets **vLLM 0.21.0** only in Phase 1 |
| Async scheduling | Must disable when R3 enabled |
| HTTP payload | Nested int lists, not base64 npy |
| Expert parallel | Opt-in via `--vllm-enable-expert-parallel` |
| Runtime patch | Lost on container rebuild; bake into Docker for production |
| Colocate CI | Reference is SGLang today; vLLM port of `test_qwen3_30B_A3B_r3.py` is pending |

---

## Phase 2 — Upgrade to Latest vLLM (follow-up issue)

1. Diff upstream `/inference/v1/generate` schema vs 0.21.0 patch.
2. Re-run Gates A → B → C.
3. Revisit async scheduling + EP launch defaults.
4. Drop or shrink `vllm.patch` if upstream exposes routing natively.
5. Extend `test_qwen3_30B_A3B_r3.py` vLLM variant to run in CI against pinned latest version.

---

## Reviewer Checklist

- [ ] Phase 1 scope clearly limited to vLLM 0.21.0
- [ ] vLLM patch reviewed (`protocol.py`, `serving.py`)
- [ ] `vllm_rollout` merge logic matches Megatron contract
- [ ] `vllm_engine` R3 launch flags documented
- [ ] **Gate B (decoupled)** passed
- [ ] **Gate C (colocate)** passed — aligned with `test_qwen3_30B_A3B_r3.py`
- [ ] Phase 2 upgrade issue filed

**Comment by CalvinXKY** [comment]
<img width="1364" height="1170" alt="Image" src="https://github.com/user-attachments/assets/a8210cf9-0869-4d96-a2ea-1997e58f6966" />

**Comment by aoshen02** [comment]
https://github.com/vllm-project/vllm/pull/38939/changes#diff-4337bfd2b38c2b70b279e8ad2d4afd6a9fe684b40cd3bfe4c6d6da8b6f261a17 v1 completions
